File: 0003_Longest_Substring_Without_Repeating_Characters/main.cpp
// T = O(n), S = O(m)
class Solution {
  public:
    int lengthOfLongestSubstring(string s) {
        int res = 0;
        unordered_map<char, int> map;
        int i = 0, j = 0;
        for (; i < s.size(); i++) {
            for (; j < s.size(); j++) {
                if (map[s[j]] == 1) {
                    break;
                }
                map[s[j]]++;
            }
            res = max(res, j - i);
            map[s[i]]--;
        }
        return res;
    }
};

// T = O(n), S = O(m)
class Solution {
  public:
    int lengthOfLongestSubstring(string s) {
        int res = 0;
        vector<int> index(255, -1);
        int i = 0, j = 0;
        for (; j < s.size(); j++) {
            i = max(index[s[j]] + 1, i);
            res = max(res, j - i + 1);
            index[s[j]] = j;
        }
        return res;
    }
};

File: 0003_Longest_Substring_Without_Repeating_Characters/golang/main.go
func max(a int, b int) int {
	if a >= b {
		return a
	}
	return b
}

func lengthOfLongestSubstring(s string) int {
	if len(s) == 0 {
		return 0
	}
	n := len(s)
	mMap := make(map[byte]int)
	res := 0
	i := 0
	j := 0
	for j < n {
		mMap[s[j]]++
		for i < j && mMap[s[j]] > 1 {
			mMap[s[i]]--
			i++
		}
		res = max(res, j-i+1)
		j++
	}
	return res
}